Add setting to hide menu bar

connection-reset opened the pull request:
<!-- You can remove this first section if you have contributed before -->
### First time contributor checklist
<!-- replace the empty checkboxes [ ] below with checked ones [x] accordingly -->
- [x] I have read the [README](https://github.com/WhisperSystems/Signal-Desktop/blob/master/README.md) and [Contributor Guidelines](https://github.com/WhisperSystems/Signal-Desktop/blob/master/CONTRIBUTING.md)
- [x] I have signed the [Contributor Licence Agreement](https://whispersystems.org/cla/)

### Contributor checklist
<!-- replace the empty checkboxes [ ] below with checked ones [x] accordingly -->
- [x] My contribution is fully baked and ready to be merged as is
- [x] My changes are rebased on the latest master branch
- [x] My commits are in nice logical chunks
- [x] I have followed the [best practices](http://chris.beams.io/posts/git-commit/) in my commit messages
- [x] I have made the choice whether I want the [BitHub reward](https://github.com/WhisperSystems/Signal-Android/wiki/BitHub-Rewards) or not by omitting or adding the word `FREEBIE` in my Git commit messages
- [ ] I have tested my contribution on these platforms:
 * GNU Hurd 1.0, Chrom{e,ium} X.Y.Z
 * Amiga OS 3.1, Chrom{e,ium} Z.Y
- [ ] My changes pass all the [local tests](https://github.com/WhisperSystems/Signal-Desktop/blob/master/CONTRIBUTING.md#tests) 100%
- [x] I have considered whether my changes need additional [tests](https://github.com/WhisperSystems/Signal-Desktop/blob/master/CONTRIBUTING.md#tests), and in the case they do, I have written them

----------------------------------------

### Description
<!--
Describe briefly what your pull request proposes to fix. Especially if you have more than one commit, it is helpful to give a summary of what your contribution as a whole is trying to solve.
Also, please describe shortly how you tested that your fix actually works.

Please note, that after you have submitted your PR, the Travis CI build check will fail. This is a known issue (#708) and you don't have to worry about it. (■_■¬)
-->
Add a setting in the themes section to hide the menu bar.
The menu bar is not needed in everyday use and might not fit in with signals dark theme.
The hidden menu bar can still be shown by pressing alt.
This basically exposes [win.setAutoHideMenuBar(hide)](https://electron.atom.io/docs/api/browser-window/#winsetautohidemenubarhide) to the user.

I did not check pass all the local tests because upstream master does not pass.
I get one failing test in spellChecker, that might be an issue with my computer.

![2017-10-10_14-33-31](https://user-images.githubusercontent.com/6831775/31386721-1c162d6e-adc8-11e7-9ec3-200854fc9e51.png)

liliakai commented on js/views/settings_view.js:
```diff
@@ -18,6 +18,7 @@
             var value = e.target.checked;
             storage.put(this.name, value);
             console.log(this.name, 'changed to', value);
+            this.$el.trigger('change-theme');
```
Eh?

liliakai replied:
Eh? Why?

liliakai commented on js/views/app_view.js:
```diff
@@ -16,6 +16,9 @@
         applyTheme: function() {
+            var hideMenuBar = storage.get('hide-menu-bar', false);
+            window.setAutoHideMenuBar(hideMenuBar);
+            window.setMenuBarVisibility(!hideMenuBar);
```
Oh I see, applyTheme is now more like applyAllTheThings ?

connection-reset replied:
I found menu bar and theme close enough for it to be in the same function.
Do you want me to move it?

scottnonnenberg replied:
I think a separate event would be good. Something like 'change-menu' instead of 'change-theme.'

connection-reset commented:
I made a 'change-hide-menu' event and corresponding function in app_view.
I also made it so that 'change-hide-menu' and 'change-theme' are only triggered when their setting changes.
Previously the notification settings would fire the events too since they use the same view and the event to fire was hardcoded in the change listener.

scottnonnenberg commented:
I just tried it and it works great - thanks! It even comes back when I restart the app. :0)

Though, much like the white loading screen when you have the dark theme enabled, the menu bar always appears on launch, and then only later goes away. I looked at [the docs](https://github.com/electron/electron/blob/master/docs/api/browser-window.md), and it seems like you can provide `autoHideMenuBar` on initial creation of a `BrowserWindow`, but `menuBarVisibility` is not an option. Would have been nice to modify `captureAndSaveWindowStats()` in `main.js` to include this, so we got the right behavior immediately on startup. Maybe you could take a quick look at that?

The other thing I wanted to talk about is the test failure you're seeing. We run tests on Linux and Windows on every push with our CI systems, so I was surprised to see that master is failing for you. Could you provide that error? What's your OS/Version?

connection-reset commented:
I added `autoHideMenuBar` to the BrowserWindow config. When setting `autoHideMenuBar` to `true` on start up, the app starts with the menu bar hidden.
`isMenuBarAutoHide()` is independent of menu bar visibility so it does the right thing (`return true`) when the menu bar is temporarily shown by pressing alt.
It would not create a config.json for me because I never move or resize the window so I changed it to save on close. That event happens before the window is destroyed so position etc is still there.

[Here is the output of `grunt unit-tests`](https://gist.github.com/connection-reset/8c97090a13a011e3c7fc56c0dbf366c4).
I run up-to-date Arch Linux with node lts v6.11.3 and hunspell 1.6.2. Hunspell is configured with english and german dictionaries and works in LibreOffice.
This happens on my laptop and on my desktop computer, but they run the same software and are really similar set up.

scottnonnenberg commented:
Glad to see we can remove that menu bar on startup! However, I think we want to save the settings on all three events: move, resize and close. I think we're ready to merge this once you've added those two previous events back.

You know what? I think the test failure is because the tests assume the current locale is English, or at least a language that says 'correct' is spelled correctly, and 'fhqwgads' is not spelled correctly. I'll have to think about the right way to test the spell-checker for multiple locales...

connection-reset commented:
I have restored the resize and move event listeners, squashed everything into 3 distinct commits and rebased on master.

You are right, both my computers are running with german locale.
```
LANG=en_US grunt unit-tests
Running "unit-tests" task
WARNING: the "timeoutsAsyncScript" command will be depcrecated soon. Please use a different command in order to avoid failures in your test after updating WebdriverIO.
>> 213 tests passed.
```
```
LANG=de_DE grunt unit-tests
...
Fatal error: Found 1 failing unit tests.
```

MBLHarrison commented:
IF you hide the menu bar, is it possible to access Preferences again?

gasi-signal commented:
@MBLHarrison On Windows, you can press <kbd>Alt</kbd> to temporarily show the menu bar and on Linux, you can try <kbd>Alt</kbd>+<kbd>F</kbd> to show the <kbd>F</kbd>ile menu, I believe.
